x/crosschain/keeper/grpc_query_cctx_rate_limit_test.go

```diff
@@ -123,60 +124,486 @@ func setupForeignCoins(
+			ethMindedCctxs:  ethMindedCctxs,
+			ethPendingCctxs: ethPendingCctxs,
+			ethPendingNonces: observertypes.PendingNonces{
+				ChainId:   ethChainID,
+				NonceLow:  1099,
+				NonceHigh: 1199,
+				Tss:       tss.TssPubkey,
+			},
+			btcMinedCctxs:   btcMinedCctxs,
+			btcPendingCctxs: btcPendingCctxs,
+			btcPendingNonces: observertypes.PendingNonces{
+				ChainId:   btcChainID,
+				NonceLow:  1099,
+				NonceHigh: 1199,
+				Tss:       tss.TssPubkey,
+			},
+			currentHeight:        1199,
+			queryLimit:           keeper.MaxPendingCctxs,
+			expectedCctxs:        append(append([]*types.CrossChainTx{}, ethPendingCctxs...), btcPendingCctxs...),
+			expectedTotalPending: 400,
+			expectedTotalValue:   3450, // 1150 (window) * (2.5 + 0.5)
+			rateLimitExceeded:    false,
+		},
+		{
+			name:            "set lower request limit to early break the LoopForwards loop",
+			rateLimitFlags:  createTestRateLimiterFlags(500, math.NewUint(5000), zrc20ETH, zrc20BTC, zrc20USDT, "2500", "50000", "0.8"),
+			ethMindedCctxs:  ethMindedCctxs,
+			ethPendingCctxs: ethPendingCctxs,
+			ethPendingNonces: observertypes.PendingNonces{
+				ChainId:   ethChainID,
+				NonceLow:  1099,
+				NonceHigh: 1199,
+				Tss:       tss.TssPubkey,
+			},
+			btcMinedCctxs:   btcMinedCctxs,
+			btcPendingCctxs: btcPendingCctxs,
+			btcPendingNonces: observertypes.PendingNonces{
+				ChainId:   btcChainID,
+				NonceLow:  1099,
+				NonceHigh: 1199,
+				Tss:       tss.TssPubkey,
+			},
+			currentHeight:        1199,
+			queryLimit:           300, // 300 < keeper.MaxPendingCctxs
+			expectedCctxs:        append(append([]*types.CrossChainTx{}, ethPendingCctxs[0:100]...), btcPendingCctxs...),
+			expectedTotalPending: 400,
+			expectedTotalValue:   1250, // 500 * 0.5 + 400 * 2.5
+			rateLimitExceeded:    false,
+		},
+		{
+			name:            "set rate to middle value (1200 < rate < 1500) to early break the LoopForwards loop with criteria #2",
+			rateLimitFlags:  createTestRateLimiterFlags(500, math.NewUint(1300), zrc20ETH, zrc20BTC, zrc20USDT, "2500", "50000", "0.8"), // 1200 < 1300 < 1500
+			ethMindedCctxs:  ethMindedCctxs,
+			ethPendingCctxs: ethPendingCctxs,
+			ethPendingNonces: observertypes.PendingNonces{
+				ChainId:   ethChainID,
+				NonceLow:  1099,
+				NonceHigh: 1199,
+				Tss:       tss.TssPubkey,
+			},
+			btcMinedCctxs:   btcMinedCctxs,
+			btcPendingCctxs: btcPendingCctxs,
+			btcPendingNonces: observertypes.PendingNonces{
+				ChainId:   btcChainID,
+				NonceLow:  1099,
+				NonceHigh: 1199,
+				Tss:       tss.TssPubkey,
+			},
+			currentHeight: 1199,
+			queryLimit:    keeper.MaxPendingCctxs,
+			// 120 ETH cctx + 200 BTC cctx
+			expectedCctxs:        append(append([]*types.CrossChainTx{}, ethPendingCctxs[0:120]...), btcPendingCctxs...),
+			expectedTotalPending: 400,
+			rateLimitExceeded:    true,
+		},
+		{
+			name: "set low rate and narrow window to early break the LoopForwards loop with criteria #2",
```

Comment:
Same, what is criteria 2?
Is it the comment below?

Reply:
https://github.com/zeta-chain/node/pull/2060/commits/578afccb54c86839c9c505dab05893132164b7ee